### PR TITLE
APIKIT-2588: Return Bad Request when malformed URI params found

### PR DESCRIPTION
**Pull request by JuanAller**

### CODE REVIEW

File: src/main/java/org/mule/module/apikit/uri/URICoder.java [hunk 1]
```diff
@@ -82,6 +102,24 @@ public static String encode(String s, Set<Character> chars) {
     return ascii ? encode_ASCII(s, chars) : encode_UTF8(s, chars);
   }
 
+  /**
+   * @param path  The path to check if contains special characters
+   * @param skipChars An ASCII set of characters that should not be considered if found in the string.
+   * @return If the path contains special characters not encoded
+   */
+  private static boolean containsSpecialChars(String path, Set<Character> skipChars) {
+    if (path.length() == 0 || path == null) {
```
Comment (juanchib) [inline src/main/java/org/mule/module/apikit/uri/URICoder.java:111]: NPE ?

File: src/test/java/org/mule/module/apikit/UrlUtilsTestCase.java [hunk 2]
```diff
@@ -0,0 +1,27 @@
...
+ */
+package org.mule.module.apikit;
+
+import static org.hamcrest.MatcherAssert.assertThat;
+import static org.hamcrest.core.IsEqual.equalTo;
+
+import org.junit.Test;
+import org.mule.module.apikit.api.UrlUtils;
+
+public class UrlUtilsTestCase {
```
Comment (dcourtil) [inline src/test/java/org/mule/module/apikit/UrlUtilsTestCase.java:15]: Could you please add a test where the base path has no `/` character?

File: src/main/java/org/mule/module/apikit/Router.java [hunk 3]
```diff
@@ -179,15 +182,15 @@ public ApiSpecification getRaml() {
...
-    String rawRequestPath = isEncoded ? attributes.getRawRequestPath() : UrlUtils.encode(attributes.getRawRequestPath());
-    String path = UrlUtils.getRelativePath(attributes.getListenerPath(), rawRequestPath);
-    return path.isEmpty() ? "/" : path;
+  private String getRequestPath(HttpRequestAttributes attributes) throws BadRequestException {
+    try {
+      String rawRequestPath = URICoder.encodeRequestPath(attributes.getRawRequestPath());
+      String path = UrlUtils.getRelativePath(attributes.getListenerPath(), rawRequestPath);
+      return path.isEmpty() ? "/" : path;
+    } catch (Exception e) {
+      throw new InvalidUriParameterException(e.getMessage());
```
Comment (dcourtil) [inline src/main/java/org/mule/module/apikit/Router.java:191]: Isn't worth to add a simple MUnit test for asserting this scenario?

File: src/main/java/org/mule/module/apikit/uri/URICoder.java (hunk 1 above)
Comment (JuanAller) [reply src/main/java/org/mule/module/apikit/uri/URICoder.java:111]: Sure!!!

File: src/main/java/org/mule/module/apikit/uri/URICoder.java [hunk 4]
```diff
@@ -56,13 +56,33 @@
...
 
   // Encoder
   // ==========================================================================
 
+  /**
+   * @param rawRequestPath  request path with original encoding /api/myEndpoint
+   * @return @rawRequestPath with encoded special characters if found
+   * @throws RuntimeException when bad encoding detected
+   */
+  public static String encodeRequestPath(String rawRequestPath) throws RuntimeException {
```
[the remark this answers is not in the document]
Comment (JuanAller) [reply src/main/java/org/mule/module/apikit/uri/URICoder.java:78]: yes sir!

File: src/test/java/org/mule/module/apikit/UrlUtilsTestCase.java (hunk 2 above)
Comment (JuanAller) [reply src/test/java/org/mule/module/apikit/UrlUtilsTestCase.java:15]: I think this is not a real case scenario using http, I will give it a try in Studio

File: src/main/java/org/mule/module/apikit/Router.java (hunk 3 above)
Comment (JuanAller) [reply src/main/java/org/mule/module/apikit/Router.java:191]: Sure! will be addeed

File: src/test/java/org/mule/module/apikit/UrlUtilsTestCase.java (hunk 2 above)
Comment (JuanAller) [reply src/test/java/org/mule/module/apikit/UrlUtilsTestCase.java:15]: check the last assertion assertThat(UrlUtils.getRelativePath("api", "/api"), equalTo("/"));

Comment (dcourtil) [reply src/test/java/org/mule/module/apikit/UrlUtilsTestCase.java:15]: yeah, right, I was just thinking about what the method was doing... looks good!

**Pull request merged**